contracts/twap/CumulativeTwap.sol

```diff
@@ -11,148 +10,208 @@ contract CumulativeTwap is BlockContext {
-    Observation[256] public observations;
+    // if the price is updated every 2 secs, 1hr twap Observation should have 60 / 2 * 60 = 1800 slots
+    Observation[MAX_OBSERVATION] public observations;
 
-    uint8 public currentObservationIndex;
+    //
+    // INTERNAL
+    //
 
-    function _update(uint256 price, uint256 lastUpdatedTimestamp) internal {
-        // for the first time update
+    function _update(uint256 price, uint256 lastUpdatedTimestamp) internal returns (bool) {
+        // for the first time updating
         if (currentObservationIndex == 0 && observations[0].timestamp == 0) {
             observations[0] = Observation({ price: price, priceCumulative: 0, timestamp: lastUpdatedTimestamp });
-            emit PriceUpdated(price, lastUpdatedTimestamp, 0);
-            return;
+            return true;
         }
 
-        // CT_IT: invalid timestamp
         Observation memory lastObservation = observations[currentObservationIndex];
-        require(lastUpdatedTimestamp > lastObservation.timestamp, "CT_IT");
 
-        // overflow of currentObservationIndex is desired since currentObservationIndex is uint8 (0 - 255),
-        // so 255 + 1 will be 0
-        currentObservationIndex++;
+        // CT_IT: invalid timestamp
+        require(lastUpdatedTimestamp >= lastObservation.timestamp, "CT_IT");
+
+        // DO NOT accept same timestamp and different price
+        // CT_IPWU: invalid price when update
+        if (lastUpdatedTimestamp == lastObservation.timestamp) {
+            require(price == lastObservation.price, "CT_IPWU");
+        }
 
-        uint256 elapsedTime = lastUpdatedTimestamp - lastObservation.timestamp;
+        // if the price remains still, there's no need for update
+        if (price == lastObservation.price) {
+            return false;
+        }
+
+        // ring buffer index, make sure the currentObservationIndex is less than MAX_OBSERVATION
+        currentObservationIndex = (currentObservationIndex + 1) % MAX_OBSERVATION;
+
+        uint256 timestampDiff = lastUpdatedTimestamp - lastObservation.timestamp;
         observations[currentObservationIndex] = Observation({
-            priceCumulative: lastObservation.priceCumulative + (lastObservation.price * elapsedTime),
+            priceCumulative: lastObservation.priceCumulative + (lastObservation.price * timestampDiff),
             timestamp: lastUpdatedTimestamp,
             price: price
         });
-
-        emit PriceUpdated(price, lastUpdatedTimestamp, currentObservationIndex);
+        return true;
     }
 
-    function _calculateTwapPrice(
+    function _calculateTwap(
         uint256 interval,
-        uint256 latestPrice,
+        uint256 price,
         uint256 latestUpdatedTimestamp
     ) internal view returns (uint256) {
+        // for the first time calculating
+        if ((currentObservationIndex == 0 && observations[0].timestamp == 0) || interval == 0) {
```

Comment:
Yes, we can. It depends on how we define the spec. 

The reason I put here is that 0 interval seems meaningless to this function.

For returning 0 case, this means meaningless for this function. Right now we have:
1. Not enough historical data (0 observation):
`(currentObservationIndex == 0 && observations[0].timestamp == 0) `
2. Not enough historical data (not enough observation): 
`else if (beforeOrAt.timestamp == atOrAfter.timestamp) `
3. 0 interval, `interval == 0`


!3. is debatable. we can return price actually. For 1. and 2., I think 0 makes sense. 

Any thoughts?

Reply:
For these 3 cases, I prefer to not return `0` or all return `0`. So maybe we don't change it and I'll just add some comments for it.

Reply:
lgtm thanks

Reply:
Just add some comments for it: https://github.com/perpetual-protocol/perp-oracle-contract/pull/76
Also add `remappings.txt` so that solidity plugins can work with foundry in editors.

Reply:
updated at f09bf06